<a href="https://colab.research.google.com/github/OnsongoMabeya/Implementation_of_descision_trees/blob/main/Week_8_Decision_tree.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Decision Tree

$entropy = \sum_i p_i log_2 p_i $

In [ ]:
import numpy as np

In [ ]:
(-10/10 * np.log2(10/10))

-0.0

- for the above it is 10 ladies and 0 men
 - entropy is 0
---
- what if it was 5 ladies and 5 men

In [ ]:
(-5/10 * np.log2(5/10)) + (-5/10 * np.log2(5/10))

1.0

- then the maxumum entropy is 1

---
- entropy can be used to know who pure or impure
 - the more the entropy, the more impure it is

---
- information gain tells us how important an information is

---
-  now let us calculaate entropy in the left child

In [ ]:
(-2/6 * np.log2(2/6)) + (-4/6 * np.log2(4/6))

0.9182958340544896

---
-  now let us calculate entropy in the right child

In [ ]:
(-3/4 * np.log2(3/4)) + (-1/4 * np.log2(1/4))

0.8112781244591328

---
- gettin the weighted entropy
 - comparing with what came from parent
 - Weighted entropy of left child

In [ ]:
(6/10 * 0.9182958340544896)

0.5509775004326937

 - Weighted entropy of right child

In [ ]:
(4/10 * 0.8112781244591328)

0.32451124978365314

 - Weighted entropy of both

In [ ]:
0.5509775004326937 + 0.32451124978365314

0.8754887502163469

- minus from the parent entropy

In [ ]:
1 - 0.5509775004326937 - 0.32451124978365314

0.12451124978365319

- information gain is: 0.12451124978365319

---
- using the information of information gain, we can tell the most important gain

In [ ]:
(-2/4 * np.log2(2/4)) + (-2/4 * np.log2(2/4))

1.0

In [ ]:
# Calculate for x

In [ ]:
(-2/4 * np.log2(2/4)) + (-2/4 * np.log2(2/4)) # root

1.0

In [ ]:
(-2/3 * np.log2(2/3)) + (-1/3 * np.log2(1/3)) # left

0.9182958340544896

In [ ]:
(-1/1 * np.log2(1/1)) # right

-0.0

In [ ]:
# Calculate for y

In [ ]:
(-2/4 * np.log2(2/4)) + (-2/4 * np.log2(2/4)) # root

1.0

In [ ]:
(-2/2 * np.log2(2/2)) # left

-0.0

In [ ]:
(-2/2 * np.log2(2/2)) # right

-0.0

In [ ]:
1 - (2/4 * 0) - (2/4 * 0) # weighted entropy if split is at y

1.0

In [ ]:
# Calculate for z

In [ ]:
(-2/4 * np.log2(2/4)) + (-2/4 * np.log2(2/4)) # root

1.0

In [ ]:
(-1/2 * np.log2(1/2)) + (-1/2 * np.log2(1/2)) # left

1.0

In [ ]:
(-1/2 * np.log2(1/2)) + (-1/2 * np.log2(1/2)) # left

1.0

In [ ]:
1 - (1/2 * 1) - (1/2 * 1) # weighted entropy if split is at y

0.0

In [ ]:
# We take y as it has the hughest entropy

# Implementation of descision trees

---


In [1]:
import pandas as pd

---
- we shall need some data

In [4]:
data=pd.read_csv("/content/vertebrate.csv", header='infer')

In [5]:
data

Name  Warm-blooded  Gives Birth  Aquatic Creature  \
0           human             1            1                 0   
1          python             0            0                 0   
2          salmon             0            0                 1   
3           whale             1            1                 1   
4            frog             0            0                 1   
5          komodo             0            0                 0   
6             bat             1            1                 0   
7          pigeon             1            0                 0   
8             cat             1            1                 0   
9   leopard shark             0            1                 1   
10         turtle             0            0                 1   
11        penguin             1            0                 1   
12      porcupine             1            1                 0   
13            eel             0            0                 1   
14     salamander             0            0                 1   

    Aerial Creature  Has Legs  Hibernates       Class  
0                 0         1           0     mammals  
1                 0         0           1    reptiles  
2                 0         0           0      fishes  
3                 0         0           0     mammals  
4                 0         1           1  amphibians  
5                 0         1           0    reptiles  
6                 1         1           1     mammals  
7                 1         1           0       birds  
8                 0         1           0     mammals  
9                 0         0           0      fishes  
10                0         1           0    reptiles  
11                0         1           0       birds  
12                0         1           1     mammals  
13                0         0           0      fishes  
14                0         1           1  amphibians